[SPARK-29523][SQL] SHOW COLUMNS should do multi-catalog resolution.

planga82 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Add ShowColumnsStatement and make SHOW COLUMNS go through the same catalog/table resolution framework of v2 commands.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
It's important to make all the commands have the same table resolution behavior, to avoid confusing end-users. e.g.

USE my_catalog
DESC t // success and describe the table t from my_catalog
SHOW COLUMNS FROM t // report table not found as there is no table t in the session catalog

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
yes. When running SHOW COLUMNS Spark fails the command if the current catalog is set to a v2 catalog, or the table name specified a v2 catalog.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Unit tests.

AmplabJenkins commented:
Can one of the admins verify this patch?

planga82 commented:
cc: @cloud-fan @rdblue @viirya @imback82

AmplabJenkins commented:
Can one of the admins verify this patch?

viirya commented:
ok to test

SparkQA commented:
**[Test build #112346 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112346/testReport)** for PR 26182 at commit [`52338e6`](https://github.com/apache/spark/commit/52338e66684d10f205b092612b15ae7166f6945e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17326/
Test PASSed.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2728,4 +2728,19 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+   *
+   * The syntax of using this command in SQL is:
+   * {{{
+   *   SHOW COLUMNS (FROM | IN) table_identifier [(FROM | IN) database];
```
table_identifier -> multi_part_name

planga82 replied:
Yes, copy / paste problem

SparkQA commented:
**[Test build #112346 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112346/testReport)** for PR 26182 at commit [`52338e6`](https://github.com/apache/spark/commit/52338e66684d10f205b092612b15ae7166f6945e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112346/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17329/
Test PASSed.

SparkQA commented:
**[Test build #112350 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112350/testReport)** for PR 26182 at commit [`81e24d5`](https://github.com/apache/spark/commit/81e24d54a5a6f4c59e6f6ecae02cf92469ef6797).

SparkQA commented:
**[Test build #112350 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112350/testReport)** for PR 26182 at commit [`81e24d5`](https://github.com/apache/spark/commit/81e24d54a5a6f4c59e6f6ecae02cf92469ef6797).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112350/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2728,4 +2728,19 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+   *
+   * The syntax of using this command in SQL is:
+   * {{{
+   *   SHOW COLUMNS (FROM | IN) multi_part_name [(FROM | IN) database];
```
The syntax should be ` SHOW COLUMNS [(FROM | IN) multi_part_name]`

cloud-fan commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -192,7 +192,7 @@ statement
         ('(' key=tablePropertyKey ')')?                                #showTblProperties
-    | SHOW COLUMNS (FROM | IN) tableIdentifier
+    | SHOW COLUMNS (FROM | IN) multipartIdentifier
         ((FROM | IN) db=errorCapturingIdentifier)?                     #showColumns
```
https://mariadb.com/kb/en/library/show-columns/

It seems like when `(FROM | IN) db` is specific, the table name must be a single part. Shall we enforce it as well?

cloud-fan replied:
BTW, in v2 command, we use `namespace` to replace `database`, as `namespace` is a more general concept. The name of `namespace` is also multi-part, so we should use `multipartIdentifier`.

planga82 replied:
Thank you for the explanation, now I have more clear some concepts. 
I agree with the proposal of maintaining the table name as a single part when namespace is defined.

huaxingao commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DDLParserSuite.scala:
```diff
@@ -961,6 +961,29 @@ class DDLParserSuite extends AnalysisTest {
+    val sql2 = "SHOW COLUMNS IN db1.t1"
+    val sql3 = "SHOW COLUMNS FROM t1 IN db1"
+    val sql4 = "SHOW COLUMNS FROM db1.t1 IN db2"
+
```
add a case for  ```a.b.c```?

planga82 replied:
Ok, I'm going to include it

SparkQA commented:
**[Test build #112618 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112618/testReport)** for PR 26182 at commit [`18871cb`](https://github.com/apache/spark/commit/18871cb57d3d25c2524c032549cf727acd8da4b4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17583/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4:
```diff
@@ -196,8 +196,10 @@ statement
+    | SHOW COLUMNS (FROM | IN)
+        ((tableName=errorCapturingIdentifier
+          (FROM | IN) namespace=multipartIdentifier) |
+        table=multipartIdentifier)                                     #showColumns
```
So this works, but makes the syntax rule a little bit complicated. Another approach is
```
SHOW COLUMNS (FROM | IN) tableName= multipartIdentifier
    ((FROM | IN) namespace=multipartIdentifier)?
```

In `visitShowColumns`, if `namespace` is specified, we check the length of `tableName`, and throw exception if it has more than one part.

By doing this, we also have a chance to give better error message, telling users that they can only use single-part table name if `FROM namespace` is specified.

planga82 replied:
Ok, I was not sure about the error message, it was not very intuitive. It is better as the way you say me.

planga82 replied:
@cloud-fan 
I reviewed the test cases and code and I arrived the conclusion that something like that is permitted:
SHOW COLUMNS FROM db1.t1 IN db1
So I have change this pull request to allow it and don't break tests in other parts.
Are you agree with it?
Thanks

SparkQA commented:
**[Test build #112693 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112693/testReport)** for PR 26182 at commit [`5ca6290`](https://github.com/apache/spark/commit/5ca62902bf06df119ce2ff0666a51e61ce189703).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17647/
Test PASSed.

SparkQA commented:
**[Test build #112693 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112693/testReport)** for PR 26182 at commit [`5ca6290`](https://github.com/apache/spark/commit/5ca62902bf06df119ce2ff0666a51e61ce189703).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowColumnsStatement(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112693/
Test FAILed.

SparkQA commented:
**[Test build #112709 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112709/testReport)** for PR 26182 at commit [`4b93dfd`](https://github.com/apache/spark/commit/4b93dfd015d7b75bf8ace32b5018abdc54d8851f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17657/
Test PASSed.

SparkQA commented:
**[Test build #112709 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112709/testReport)** for PR 26182 at commit [`4b93dfd`](https://github.com/apache/spark/commit/4b93dfd015d7b75bf8ace32b5018abdc54d8851f).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112709/
Test FAILed.

planga82 commented:
It seems unrelated error in the tests.

maropu commented:
@planga82 you can re-run the Jenkins tests by yourself

maropu commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #112715 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112715/testReport)** for PR 26182 at commit [`4b93dfd`](https://github.com/apache/spark/commit/4b93dfd015d7b75bf8ace32b5018abdc54d8851f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17660/
Test PASSed.

SparkQA commented:
**[Test build #112715 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112715/testReport)** for PR 26182 at commit [`4b93dfd`](https://github.com/apache/spark/commit/4b93dfd015d7b75bf8ace32b5018abdc54d8851f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112715/
Test FAILed.

SparkQA commented:
**[Test build #112718 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112718/testReport)** for PR 26182 at commit [`56c6847`](https://github.com/apache/spark/commit/56c6847f2e05dfa0a97737cb51e4954be1a678a7).

planga82 commented on sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala:
```diff
@@ -2078,23 +2078,6 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
     }
   }
 
-  test("show columns - negative test") {
```
This test now has nos sense, because it assumes that you can put a multipart table identifier with a namespace defined

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17663/
Test PASSed.

SparkQA commented:
**[Test build #112718 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112718/testReport)** for PR 26182 at commit [`56c6847`](https://github.com/apache/spark/commit/56c6847f2e05dfa0a97737cb51e4954be1a678a7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112718/
Test FAILed.

planga82 commented:
I am reviewing it

planga82 committed:
[SPARk-29523][SQL] show columns multi-catalog resolution

[SPARR-29523] Update show columns namespace concept


[SPARK-2953] Fix rebase problemas


New impl

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17671/
Test PASSed.

SparkQA commented:
**[Test build #112726 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112726/testReport)** for PR 26182 at commit [`461bf15`](https://github.com/apache/spark/commit/461bf15960b1d9bbd48e92712070d169a342868f).

SparkQA commented:
**[Test build #112726 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112726/testReport)** for PR 26182 at commit [`461bf15`](https://github.com/apache/spark/commit/461bf15960b1d9bbd48e92712070d169a342868f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowColumnsStatement(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112726/
Test FAILed.

imback82 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2877,4 +2877,28 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+    } else {
+      ShowColumnsStatement(table, None)
+    }
+  }
```
You can simplify the code a bit if you want:
```scala
    val table = visitMultipartIdentifier(ctx.tableName)
    val namespace = Option(ctx.multipartIdentifier).map(visitMultipartIdentifier)
    if (namespace.isDefined && namespace.get.length > 1 && table.length > 1) {
      throw new ParseException(
        s"If namespace is defined, table can not be multipart: ${table.mkString(".")}", ctx)
    }
    ShowColumnsStatement(table, namespace)
```

planga82 replied:
Elegant solution, thank you!!

imback82 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2877,4 +2877,28 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+      val namespace = visitMultipartIdentifier(ctx.namespace)
+      if (table.length > 1 && namespace.length > 1) {
+        throw new ParseException(
+          s"If namespace is defined, table can not be multipart: ${table.mkString(".")}", ctx)
```
let's use `quoted` instead of mkString since it's multi part identifier.

imback82 replied:
Also, the exception message is different from what is actually checked (defined AND namespace.length > 1).

planga82 replied:
@imback82 
I have change the message with this:
"_If namespace is multipart, table can not be multipart:_"
 but i'm not sure what are you referring with quoted instead mkstring.

imback82 replied:
You can use `MultipartIdentifierHelper.quoted`. Check `visitCacheTable` for an example.

planga82 replied:
Ok, I have seen it, thank you

imback82 commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1307,6 +1307,28 @@ class DataSourceV2SQLSuite
     assert(e.message.contains(s"$sqlCommand is only supported with v1 tables"))
   }
 
+  test("SHOW COLUMNS") {
```
nit: can we move this above `testV1Command` to group v1 command tests together?

imback82 commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1307,6 +1307,28 @@ class DataSourceV2SQLSuite
   }
 
+  test("SHOW COLUMNS") {
+    val t = s"testcat.ns1.ns2.tbl"
```
nit: `s` is not needed.

SparkQA commented:
**[Test build #112803 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112803/testReport)** for PR 26182 at commit [`192bf6a`](https://github.com/apache/spark/commit/192bf6a455958ab0bb25f3722f3d711b5d715753).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17743/
Test PASSed.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2911,4 +2911,28 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+    if (namespace.isDefined && namespace.get.length > 1 && table.length > 1) {
+      val tableName = table.init
+      throw new ParseException(
+        s"If namespace is multipart, table can not be multipart: ${tableName.quoted}", ctx)
```
Just table.quoted.  visitCacheTable takes tableName.init because it only wants catalog and namespace there.

In your case, you need to show full given multipart table.

cloud-fan replied:
how about "Cannot specify multi-part table name if `FROM namespace` is specified."

viirya replied:
If the behavior is as https://github.com/apache/spark/pull/26182#discussion_r339892210, then the suggested one is more correct.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -338,6 +338,19 @@ class ResolveSessionCatalog(
+      val sql = "SHOW COLUMNS"
+      val v1TableName = parseV1Table(table, sql)
+      val db = namespace.map(parseV1Namespace(sql)).map(_.head)
+      ShowColumnsCommand(db, v1TableName.asTableIdentifier)
```
Can we just concatenate table and namespace, and pass to parseV1Table?

SparkQA commented:
**[Test build #112803 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112803/testReport)** for PR 26182 at commit [`192bf6a`](https://github.com/apache/spark/commit/192bf6a455958ab0bb25f3722f3d711b5d715753).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowColumnsStatement(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112803/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2911,4 +2911,28 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+
+    val table = visitMultipartIdentifier(ctx.table)
+    val namespace = Option(ctx.namespace).map(visitMultipartIdentifier)
+    if (namespace.isDefined && namespace.get.length > 1 && table.length > 1) {
```
I think the logic should be: if the namespace is specified, then table name length must be 1.

cloud-fan replied:
OK I get it. Previously we support `SHOW COLUMNS FROM db1.t1 IN db1`, and at runtime we check if the specified database name is the same as the db name in table identifier.

Seems like we should move this check into the analyzer, i.e. in `ResolveSessionCatalog`

cloud-fan replied:
Since we don't have v2 SHOW COLUMNS yet, so in `ResolveSessionCatalog`, we should simply check that
1. the table name has <= 2 parts
2. the namespace name should have only one part if specified

SparkQA commented:
**[Test build #112862 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112862/testReport)** for PR 26182 at commit [`6c712ef`](https://github.com/apache/spark/commit/6c712efe8632abe938f8cb1c3e1db24e9737af51).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17792/
Test PASSed.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -2912,6 +2912,32 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+    if (table.length > 2) {
+      throw new ParseException(
+        s"Table name should have at most two parts: ${table.quoted}", ctx)
+    }
```
Should we move the checks to Analyzer? Let ResolveSessionCatalog do the check?

viirya replied:
cc @cloud-fan

planga82 replied:
I put this validations here because It's sounds like sintaxis problems, and I saw other ParserExceptions in AstBuilder. In ResolveSessionCatalog I saw AnalisisExceptions that sounds more like catalog access problems. 
But is only by intuition, we can put ir wherever you want. 
Could be better to put all the validations together

cloud-fan replied:
This is a tricky problem. Ideally, we should have a check at the parser that: if namespace is specified, then table name should be a single part.

Unfortunately we support `SHOW COLUMNS FROM db1.t1 IN db1`. To keep supporting it, we can't check anything in the parser. In the analyzer, we can add a check before converting to the v1 command: if namespace is specified, it must be a single part, and the table name must have <= 2 parts.

planga82 replied:
Ok, perfect, I move this validation

SparkQA commented:
**[Test build #112862 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112862/testReport)** for PR 26182 at commit [`6c712ef`](https://github.com/apache/spark/commit/6c712efe8632abe938f8cb1c3e1db24e9737af51).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowColumnsStatement(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112862/
Test FAILed.

planga82 committed:
[SPARK-29523] Fix exception message

SparkQA commented:
**[Test build #112870 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112870/testReport)** for PR 26182 at commit [`b584047`](https://github.com/apache/spark/commit/b5840478e8a574ea3d481475f5d50b3b35bd43ce).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17798/
Test PASSed.

SparkQA commented:
**[Test build #112870 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112870/testReport)** for PR 26182 at commit [`b584047`](https://github.com/apache/spark/commit/b5840478e8a574ea3d481475f5d50b3b35bd43ce).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112870/
Test PASSed.

planga82 committed:
[SPARK-29523] Move validations

SparkQA commented:
**[Test build #112941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112941/testReport)** for PR 26182 at commit [`10d2709`](https://github.com/apache/spark/commit/10d27095d8066a3942d66f139ded54d8100cf78b).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17864/
Test PASSed.

SparkQA commented:
**[Test build #112943 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112943/testReport)** for PR 26182 at commit [`8f4cc92`](https://github.com/apache/spark/commit/8f4cc92f93044afd2a32a692dc4061a5f1c4a2e1).

imback82 commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -339,6 +339,27 @@ class ResolveSessionCatalog(
+      val sql = "SHOW COLUMNS"
+      val v1TableName = parseV1Table(table, sql).asTableIdentifier
+      val resolver = conf.resolver
+      val db = namespace match {
```
Why is this done here? Wouldn't this be done inside `ShowColumnsCommand` anyway?

planga82 replied:
I have move it because this comment. 
https://github.com/apache/spark/pull/26182#discussion_r339897559
It seems simpler to check all in one place

imback82 commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1358,6 +1358,29 @@ class DataSourceV2SQLSuite
+      }
+      assert(e1.message.contains(s"SHOW COLUMNS is only supported with v1 tables."))
+
+      val e2 = intercept[AnalysisException] {
```
You can do:
```scala
testV1Command("SHOW COLUMNS", s"FROM $t")
testV1Command("SHOW COLUMNS", s"IN $t")
```

planga82 committed:
[SPARK-29523] Fix comments

SparkQA commented:
**[Test build #112952 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112952/testReport)** for PR 26182 at commit [`5365243`](https://github.com/apache/spark/commit/5365243ee4ff41ec244d975745b0fce272a65b81).

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17874/
Test PASSed.

SparkQA commented:
**[Test build #112941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112941/testReport)** for PR 26182 at commit [`10d2709`](https://github.com/apache/spark/commit/10d27095d8066a3942d66f139ded54d8100cf78b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112941/
Test PASSed.

SparkQA commented:
**[Test build #112943 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112943/testReport)** for PR 26182 at commit [`8f4cc92`](https://github.com/apache/spark/commit/8f4cc92f93044afd2a32a692dc4061a5f1c4a2e1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112943/
Test PASSed.

SparkQA commented:
**[Test build #112952 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112952/testReport)** for PR 26182 at commit [`5365243`](https://github.com/apache/spark/commit/5365243ee4ff41ec244d975745b0fce272a65b81).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112952/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

planga82 commented:
Thanks all for your help!!
